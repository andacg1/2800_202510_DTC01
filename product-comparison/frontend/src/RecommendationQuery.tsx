import React, { useEffect, useState } from "react";
import type { Product } from "./product.ts";

type RecommendationQueryProps = {
  className?: string;
  children?: React.ReactNode;
  products?: Product[];
};

type Recommendation = {
  recommendedProductId: string;
  recommendedProductTitle: string;
  reason: string;
};

type RecommendationResponse = {
  outputJson: Recommendation;
};

const RecommendationQuery = ({
  className,
  children,
  products,
}: RecommendationQueryProps) => {
  const [response, setResponse] = useState<Recommendation>();
  const [isLoading, setIsLoading] = useState(false);
  useEffect(() => {}, []);
  // I'm a retired snowboarded who just lost everything after my 7th divorce. What kind of gear should I buy?

  return (
    <div>
      <form
        style={{
          display: "flex",
          flexDirection: "column",
          alignItems: "end",
        }}
        onSubmit={async (e) => {
          e.preventDefault();
          const data = new FormData(e.currentTarget);
          const query = data.get("query");
          // TODO: define this port in the shopify CLI config, if possible
          const url = `${process.env.APP_BACKEND_URL}/recommend`;
          console.log(url);
          if (!url) {
            return console.error("Could not find APP_BACKEND_URL");
          }
          setIsLoading(true);
          const resp = await fetch(url, {
            method: "POST",
            body: JSON.stringify({
              query,
              products,
            }),
          });
          const body: RecommendationResponse = await resp.json();
          setIsLoading(false);
          setResponse(body.outputJson);
        }}
      >
        <textarea
          className="w-full"
          placeholder={"Tell us about your use case"}
          name="query"
          style={{ width: "100%" }}
        />
        <button type="submit" style={{ width: "100px", marginTop: "8px" }}>
          Submit
        </button>
      </form>
      {isLoading ? "Loading..." : null}
      {response?.reason ? <p>{response.reason}</p> : null}
    </div>
  );
};

export default RecommendationQuery;
