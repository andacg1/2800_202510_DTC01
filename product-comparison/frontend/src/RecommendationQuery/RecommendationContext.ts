import type React from "react";
import { createContext } from "react";
import type { Product } from "../product.ts";

/**
 * Props for the RecommendationQuery component
 * @typedef {Object} RecommendationQueryProps
 * @property {string} [className] - Optional CSS class name for styling
 * @property {React.ReactNode} [children] - Optional child elements
 * @property {Product[]} [products] - Array of products to generate recommendations from
 */
export type RecommendationQueryProps = {
  className?: string;
  children?: React.ReactNode;
  products?: Product[];
};

/**
 * Structure of a product recommendation
 * @typedef {Object} Recommendation
 * @property {string} recommendedProductId - ID of the recommended product
 * @property {string} recommendedProductTitle - Title of the recommended product
 * @property {string} reason - Explanation for why this product was recommended
 */
export type Recommendation = {
  recommendedProductId: string;
  recommendedProductTitle: string;
  reason: string;
};

/**
 * Response structure from the recommendation API
 * @typedef {Object} RecommendationResponse
 * @property {Recommendation} outputJson - The recommendation data
 */
export type RecommendationResponse = {
  outputJson: Recommendation;
};

/**
 * Type definition for the Recommendation context value
 * @typedef {Object} RecommendationContextType
 * @property {Recommendation | undefined} recommendation - Current recommendation or undefined
 * @property {React.Dispatch<React.SetStateAction<Recommendation | undefined>>} setRecommendation - Function to update recommendation
 * @property {string | undefined} query - Current search query
 * @property {React.Dispatch<React.SetStateAction<string | undefined>>} setQuery - Function to update search query
 */
export type RecommendationContextType = {
  recommendation: Recommendation | undefined;
  setRecommendation: React.Dispatch<
    React.SetStateAction<Recommendation | undefined>
  >;
  query: string | undefined;
  setQuery: React.Dispatch<React.SetStateAction<string | undefined>>;
};

/**
 * Default values for the Recommendation context
 */
export const defaultRecommendationContext: RecommendationContextType = {
  recommendation: {
    recommendedProductId: "",
    recommendedProductTitle: "",
    reason: "",
  },
  setRecommendation: () => {},
  query: "",
  setQuery: () => {},
};

/**
 * React context for managing product recommendations throughout the application
 * Provides access to recommendation state and query functionality
 */
export const RecommendationContext = createContext(
  defaultRecommendationContext,
);
