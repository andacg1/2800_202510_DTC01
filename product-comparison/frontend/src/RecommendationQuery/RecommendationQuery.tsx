import React, { useContext, useEffect, useState } from "react";
import {
  defaultRecommendationContext,
  Recommendation,
  RecommendationContext,
  RecommendationQueryProps,
  RecommendationResponse,
} from "./RecommendationContext.ts";

async function mockOpenApiResponse() {
  return {
    success: true,
    outputJson: {
      recommendedProductId: "9962241655059",
      recommendedProductTitle: "Copper Light",
      reason:
        "As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the 'Copper Light' is the only item that is actually designed to produce light. Snowboards, wax, sofas, and plant pots do not provide any illumination for your underground activities. The 'Copper Light' is a bedside lamp, but it’s portable and much brighter than a typical kobold candle—making it the most logical upgrade for lighting up mine tunnels. Just make sure you find a way to power it underground!",
    },
    message: {
      id: "resp_681fca9d95d4819183c0eb7d807722830abf5240ac5e06eb",
      object: "response",
      created_at: 1746913949,
      status: "completed",
      error: null,
      incomplete_details: null,
      instructions: null,
      max_output_tokens: null,
      model: "gpt-4.1-2025-04-14",
      output: [
        {
          id: "msg_681fca9f324c8191af5b99dee4988ebe0abf5240ac5e06eb",
          type: "message",
          status: "completed",
          content: [
            {
              type: "output_text",
              annotations: [],
              text: '{"recommendedProductId":"9962241655059","recommendedProductTitle":"Copper Light","reason":"As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the \'Copper Light\' is the only item that is actually designed to produce light. Snowboards, wax, sofas, and plant pots do not provide any illumination for your underground activities. The \'Copper Light\' is a bedside lamp, but it’s portable and much brighter than a typical kobold candle—making it the most logical upgrade for lighting up mine tunnels. Just make sure you find a way to power it underground!"}',
            },
          ],
          role: "assistant",
        },
      ],
      parallel_tool_calls: true,
      previous_response_id: null,
      reasoning: {
        effort: null,
        summary: null,
      },
      service_tier: "default",
      store: true,
      temperature: 1,
      text: {
        format: {
          type: "json_schema",
          description: null,
          name: "recommendation",
          schema: {
            type: "object",
            properties: {
              recommendedProductId: { type: "string" },
              recommendedProductTitle: { type: "string" },
              reason: { type: "string" },
            },
            required: [
              "recommendedProductId",
              "recommendedProductTitle",
              "reason",
            ],
            additionalProperties: false,
          },
          strict: true,
        },
      },
      tool_choice: "auto",
      tools: [],
      top_p: 1,
      truncation: "disabled",
      usage: {
        input_tokens: 21795,
        input_tokens_details: { cached_tokens: 0 },
        output_tokens: 138,
        output_tokens_details: { reasoning_tokens: 0 },
        total_tokens: 21933,
      },
      user: null,
      metadata: {},
      output_text:
        '{"recommendedProductId":"9962241655059","recommendedProductTitle":"Copper Light","reason":"As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the \'Copper Light\' is the only item that is actually designed to produce light. Snowboards, wax, sofas, and plant pots do not provide any illumination for your underground activities. The \'Copper Light\' is a bedside lamp, but it’s portable and much brighter than a typical kobold candle—making it the most logical upgrade for lighting up mine tunnels. Just make sure you find a way to power it underground!"}',
    },
  };
}

const RecommendationQuery = ({
  className,
  children,
  products,
}: RecommendationQueryProps) => {
  const { recommendation, setRecommendation, query, setQuery } = useContext(
    RecommendationContext,
  );
  const isLoading = query && (!recommendation || !recommendation?.reason);

  return (
    <div className="mb-12">
      <form
        style={{
          display: "flex",
          flexDirection: "column",
          alignItems: "end",
        }}
        className="flex flex-col items-end mb-6"
        onSubmit={async (e) => {
          e.preventDefault();
          const queryInput = e.currentTarget.elements.namedItem(
            "query",
          ) as HTMLTextAreaElement;
          console.log(e.currentTarget.elements);
          const data = new FormData(e.currentTarget);
          const queryData = data.get("query");
          // TODO: define this port in the shopify CLI config, if possible
          if (!(typeof queryData === "string")) {
            return console.error("Query cannot be a File");
          }
          if (queryInput && queryInput?.value) {
            queryInput.value = "";
          }
          setQuery(queryData || "");
          setRecommendation(defaultRecommendationContext.recommendation);
          if (!process.env.APP_BACKEND_URL) {
            return console.error("Could not find APP_BACKEND_URL");
          }
          const url = `${process.env.APP_BACKEND_URL}/recommend`;
          console.log(url);

          let body: RecommendationResponse;

          const MOCK_OPENAI_REQUEST = false;
          if (MOCK_OPENAI_REQUEST) {
            body = await mockOpenApiResponse();
          } else {
            const resp = await fetch(url, {
              method: "POST",
              body: JSON.stringify({
                query: queryData,
                products,
              }),
            });
            body = await resp.json();
          }
          setRecommendation(body.outputJson);
        }}
      >
        <textarea
          className="textarea w-full text-2xl placeholder:opacity-40"
          placeholder={"Tell us about your use case"}
          name="query"
        />
        <button
          type="submit"
          className={"btn btn-primary"}
          style={{ width: "100px", marginTop: "8px" }}
        >
          Submit
        </button>
      </form>
      {query ? (
        <div className="chat chat-end">
          <div className="chat-bubble chat-bubble-info max-w-prose">
            {query}
          </div>
        </div>
      ) : null}
      {isLoading ? (
        <span className="loading loading-dots loading-md"></span>
      ) : null}
      {recommendation?.reason ? (
        <div className="chat chat-start mb-1">
          <div className="chat-bubble max-w-prose">
            {recommendation?.reason}
          </div>
        </div>
      ) : null}
    </div>
  );
};

export default RecommendationQuery;
