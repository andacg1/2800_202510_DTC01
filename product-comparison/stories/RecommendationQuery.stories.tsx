import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import RecommendationQuery from "../frontend/src/RecommendationQuery/RecommendationQuery";
import { RecommendationContext } from "../frontend/src/RecommendationQuery/RecommendationContext";

// Typing just enough to satisfy the component's requirements
interface SimpleProduct {
  id: number;
  title: string;
  handle: string;
  description: string;
  specs: Record<string, any>;
  [key: string]: any; // Allow other properties
}

// Sample product data for the story
const sampleProducts: SimpleProduct[] = [
  {
    id: 1,
    title: "Copper Light",
    handle: "copper-light",
    description: "A beautiful copper light for your home",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    vendor: "Lighting Co.",
    product_type: "Lighting",
    tags: ["light", "copper", "home"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 100,
    type: "Physical",
    price_min: 100,
    price_max: 100,
    available: true,
    specs: {
      color: "Copper",
      size: "Medium",
      weight: "2kg",
      available_regions: ["US", "CA", "EU"],
    },
  },
  {
    id: 2,
    title: "Snowboard Pro",
    handle: "snowboard-pro",
    description: "Professional snowboard for winter sports",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    vendor: "Snow Sports Inc.",
    product_type: "Winter Sports",
    tags: ["snow", "board", "winter"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 350,
    type: "Physical",
    price_min: 350,
    price_max: 350,
    available: true,
    specs: {
      color: "Black",
      size: "Large",
      weight: "5kg",
      available_regions: ["US", "CA", "EU"],
    },
  },
  {
    id: 3,
    title: "Plant Pot",
    handle: "plant-pot",
    description: "Ceramic pot for indoor plants",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    vendor: "Home & Garden",
    product_type: "Home Decor",
    tags: ["pot", "plant", "ceramic"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 35,
    type: "Physical",
    price_min: 35,
    price_max: 35,
    available: true,
    specs: {
      color: "White",
      size: "Small",
      weight: "1kg",
      available_regions: ["US", "EU", "JP"],
    },
  },
];

const meta = {
  title: "Product Comparison/RecommendationQuery",
  component: RecommendationQuery,
  parameters: {
    layout: "centered",
  },
  decorators: [
    (Story) => {
      // Setup process.env for the component
      if (typeof process !== "undefined") {
        process.env = {
          ...process.env,
          APP_BACKEND_URL: "https://example.com",
        };
      }

      // Create a wrapper component with the context
      const RecommendationWrapper = () => (
        <div data-theme="light">
          <RecommendationContext.Provider
            value={{
              recommendation: undefined,
              setRecommendation: () => {},
              query: undefined,
              setQuery: () => {},
            }}
          >
            <div
              style={{
                width: "800px",
                padding: "20px",
              }}
            >
              <Story />
            </div>
          </RecommendationContext.Provider>
        </div>
      );

      return <RecommendationWrapper />;
    },
  ],
  tags: ["autodocs"],
} satisfies Meta<typeof RecommendationQuery>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Empty: Story = {
  args: {
    products: [],
  },
};

export const WhileWaiting: Story = {
  args: {
    products: sampleProducts,
  },
  decorators: [
    (Story) => {
      const WithRecommendation = () => (
        <RecommendationContext.Provider
          value={{
            recommendation: undefined,
            setRecommendation: () => {},
            query:
              "I am a level 5 kobold working in dark mines. My head-candle doesn't provide enough light. Which product would help me?",
            setQuery: () => {},
          }}
        >
          <Story />
        </RecommendationContext.Provider>
      );
      return <WithRecommendation />;
    },
  ],
};

export const WithRecommendation: Story = {
  args: {
    products: sampleProducts,
  },
  decorators: [
    (Story) => {
      const WithRecommendation = () => (
        <RecommendationContext.Provider
          value={{
            recommendation: {
              recommendedProductId: "gid://shopify/Product/9962241655059",
              recommendedProductTitle: "Copper Light",
              reason:
                "As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the 'Copper Light' is the only item that is actually designed to produce light.",
            },
            setRecommendation: () => {},
            query:
              "I am a level 5 kobold working in dark mines. My head-candle doesn't provide enough light. Which product would help me?",
            setQuery: () => {},
          }}
        >
          <Story />
        </RecommendationContext.Provider>
      );
      return <WithRecommendation />;
    },
  ],
};
