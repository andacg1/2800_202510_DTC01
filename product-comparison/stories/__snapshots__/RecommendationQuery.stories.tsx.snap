// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Comparison/RecommendationQuery Empty smoke-test 1`] = `
<div data-theme="light">
  <div style="width: 800px; padding: 20px;">
    <div class="mb-12">
      <form class="flex flex-col items-end mb-6"
            style="display: flex; flex-direction: column; align-items: end;"
      >
        <textarea class="textarea w-full text-2xl placeholder:opacity-40"
                  placeholder="Tell us about your use case"
                  name="query"
        >
        </textarea>
        <button type="submit"
                class="btn btn-primary"
                style="width: 100px; margin-top: 8px;"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Product Comparison/RecommendationQuery WhileWaiting smoke-test 1`] = `
<div data-theme="light">
  <div style="width: 800px; padding: 20px;">
    <div class="mb-12">
      <form class="flex flex-col items-end mb-6"
            style="display: flex; flex-direction: column; align-items: end;"
      >
        <textarea class="textarea w-full text-2xl placeholder:opacity-40"
                  placeholder="Tell us about your use case"
                  name="query"
        >
        </textarea>
        <button type="submit"
                class="btn btn-primary"
                style="width: 100px; margin-top: 8px;"
        >
          Submit
        </button>
      </form>
      <div class="chat chat-end">
        <div class="chat-bubble chat-bubble-info max-w-prose">
          I am a level 5 kobold working in dark mines. My head-candle doesn't provide enough light. Which product would help me?
        </div>
      </div>
      <span class="loading loading-dots loading-md">
      </span>
    </div>
  </div>
</div>
`;

exports[`Product Comparison/RecommendationQuery WithRecommendation smoke-test 1`] = `
<div data-theme="light">
  <div style="width: 800px; padding: 20px;">
    <div class="mb-12">
      <form class="flex flex-col items-end mb-6"
            style="display: flex; flex-direction: column; align-items: end;"
      >
        <textarea class="textarea w-full text-2xl placeholder:opacity-40"
                  placeholder="Tell us about your use case"
                  name="query"
        >
        </textarea>
        <button type="submit"
                class="btn btn-primary"
                style="width: 100px; margin-top: 8px;"
        >
          Submit
        </button>
      </form>
      <div class="chat chat-end">
        <div class="chat-bubble chat-bubble-info max-w-prose">
          I am a level 5 kobold working in dark mines. My head-candle doesn't provide enough light. Which product would help me?
        </div>
      </div>
      <div class="chat chat-start mb-1">
        <div class="chat-bubble max-w-prose">
          As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the 'Copper Light' is the only item that is actually designed to produce light.
        </div>
      </div>
    </div>
  </div>
</div>
`;
